pkg/apis/jenkins.io/v1/types_pipelinestructure.go

```diff
@@ -0,0 +1,665 @@
+type PipelineStructureStep struct {
+	// One of command, step, or loop is required.
+	// +optional
+	Command *string `json:"command,omitempty" protobuf:"bytes,1,opt,name=command"`
+	// args is optional, but only allowed with command
+	// +optional
+	Args []string `json:"args,omitempty" protobuf:"bytes,2,opt,name=args"`
+	// dir is optional, but only allowed with command. Refers to subdirectory of workspace
+	// +optional
+	Dir *string `json:"dir,omitempty" protobuf:"bytes,3,opt,name=dir"`
+
+	// +optional
+	Step *string `json:"step,omitempty" protobuf:"bytes,4,opt,name=step"`
+	// options is optional, but only allowed with step
+	// Also, we'll need to do some magic to do type verification during translation - i.e., this step wants a number
+	// for this option, so translate the string value for that option to a number.
+	// +optional
+	Options map[string]string `json:"options,omitempty" protobuf:"bytes,5,opt,name=options"`
+
+	// +optional
+	Loop *PipelineStructureLoop `json:"loop,omitempty" protobuf:"bytes,6,opt,name=loop"`
+
+	// agent can be overridden on a step
+	// +optional
+	Agent *PipelineStructureAgent `json:"agent,omitempty" protobuf:"bytes,7,opt,name=agent"`
+}
+
+// PipelineStructureLoop is a special step that defines a variable, a list of possible values for that variable, and a set of steps to
+// repeat for each value for the variable, with the variable set with that value in the environment for the execution of
+// those steps.
+type PipelineStructureLoop struct {
+	// The variable name.
+	Variable string `json:"variable" protobuf:"bytes,1,opt,name=variable"`
+	// The list of values to iterate over
+	Values []string `json:"values" protobuf:"bytes,2,opt,name=values"`
+	// The steps to run
+	Steps []PipelineStructureStep `json:"steps" protobuf:"bytes,3,opt,name=steps"`
+}
+
+// PipelineStructureStage is a unit of work in a pipeline, corresponding either to a Task or a set of Tasks to be run sequentially or in
+// parallel with common configuration.
+type PipelineStructureStage struct {
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// +optional
+	Agent *PipelineStructureAgent `json:"agent,omitempty" protobuf:"bytes,2,opt,name=agent"`
+	// +optional
+	Options *PipelineStructureStageOptions `json:"options,omitempty" protobuf:"bytes,3,opt,name=options"`
+	// +optional
+	Environment []PipelineStructureEnvVar `json:"environment,omitempty" protobuf:"bytes,4,opt,name=environment"`
+	// +optional
+	Steps []PipelineStructureStep `json:"steps,omitempty" protobuf:"bytes,5,opt,name=steps"`
+	// +optional
+	Stages []*PipelineStructureStage `json:"stages,omitempty" protobuf:"bytes,6,opt,name=stages"`
+	// +optional
+	Parallel []*PipelineStructureStage `json:"parallel,omitempty" protobuf:"bytes,7,opt,name=parallel"`
+	// +optional
+	Post []PipelineStructurePost `json:"post,omitempty" protobuf:"bytes,8,opt,name=post"`
+	// +optional
+	Parent *string `json:"parent,omitempty" protobuf:"bytes,9,opt,name=parent"`
```

Comment:
nit : I would drop this and depth from this struct and handle all of the graph-related stuff with a separate `pipelinestructure_graph_traversal` file or similar for clear separation of concerns. We'd need pointers to the actual `PipelineStructureStages` in this struct itself to avoid having to construct a separate graph anyway, so it seems cleanest for this struct to map very closely to the YAML format itself. We can always cache the graph with its related PipelineStructure in whatever controller cares about these objects so we don't have to keep recreating it if performance is a concern. These are the kind of data structures I think we'd need for graph-related stuff:

```
Node {
    Stage           *PipelineStructureStage
    Depth           int8
    Children        []*Node
    Parent          *Node
    PreviousSibling *Node
    NextSibling     *Node
}

Graph {
    FirstStage *Node // maybe not?
    Nodes      map[*PipelineStructureStage]*Node
}
```

Reply:
Yeah, actually, we don't need `Depth` or `Parent`, but I also don't think we need anything special for graph traversal - given a stage name and a `PipelineStructure`, we can easily traverse it to find the parent/etc. I don't know if we'll ever actually have a need for full-blown graph traversal along the lines of what we have for the `FlowGraph` in Jenkins Pipeline - here we just need a way to look at a `Task` (well, technically to look at a `Pod` to get its `TaskRun` to get its `Task` and `PipelineRun` and `Pipeline`, but I'm gonna shortcut here), get its corresponding stage name, and find where it is in the `PipelineStructure` so that we can populate `PipelineActivity` correctly.